### PR TITLE
refactor: app-sync scripts

### PR DESCRIPTION
**Pull request by chrismclarke**
PR Checklist

- [x] - Latest `master` branch merged
- [x] - PR title descriptive (can be used in release notes)

## Description

Various bits of housekeeping and refactoring for app sync commands (mostly extracted or adapted from #1755)
Main changes
- Replace app-data `copy` command with separate postProcessors for sheets and assets
- Add additional QC checks for sheets and assets
- Add support for populating processed content to local deployment folder (but keep default `packages/app-data` for now)
- Change asset contents to json from .ts and separate out tracked and untracked assets (makes the app-data folder purely data and with no reliance on .ts files - this is what will eventually be synced to git repos)
- Update theme assets to remove `theme_` prefix in contents

## Review Notes
As changes are only to parsers the main check is that content synced is processed the same as previously.

The parser includes a version bump which should invalidate the cache and force all sheets to be reprocessed (around 265 sheets or so on plh global config). If this is not the case (e.g. retesting) all caches can be cleared via `yarn workflow clear_workflow_cache`. The only known content change should be the naming of theme assets which no longer have a `theme_` prefix in the contents.

Additionally, it could be useful to delete files from the `packages/app-data` folder to ensure they are repopulated as expected

It would be good to check a couple different deployments (all my testing on plh_global), as well as language and theme changes.

## Dev Notes
Changes to the copy.ts script were originally made around 10 months ago, so trying to recover now means potentially losing updates that were made since. I've manually gone through the commits and tried to apply the changes (most occurred in #1522), but there is still some scope for regression issues. Again, these should be flagged in the final data output from functional testing so probably no need for detailed code review (but good to check the functionality of themes).

## Future TODOs
- [ ] Add sheets post-process spec

## CC Checks
- [x] Review all legacy copyTS code to ensure new scripts maintain functionality (e.g. theming system updates) - can look at file history for changes post May 31st PR
- [x] Try to clean diffs for contents.json and index.ts refactor (move to separate PR)
- [x] Test local sync and running different deployments
- [x] Check if local deployment workflows still work
- [x] Fix `sync_assets` workflow
- [x] Fix `sync_sheets` workflow
- [x] Fix broken `assets.spec.ts`
- [x] Add additional asset specs for translated and theme assets
- [x] Resolve/clarify whether themes are also languages (assets quality control workaround)
- [x] Check test-visual working
- [x] Check theming system

## Git Issues

Closes #

## Screenshots/Videos

**Additional assets QC checks**
image

**Theme assets checked as working**
image

### CODE REVIEW

**Review by jfmcquade — CHANGES_REQUESTED** [review]
Looks good. 

The theme functionality (correctly handling assets when switching themes) seems to be preserved, both from inspecting the code and some functional testing. 

I flagged in the inline comments that an error is thrown when a deployment does not specify a value for `filter_language_codes`. I think this is actually indicative of a broader issue: when setting a deployment to active and compiling the `activeDeployment.json` file, default values are not set for `local_drive`, `app_data`, `translations` and `workflows`.
This means that values that are not specified in a deployment's config cannot be drawn upon correctly as they may be undefined. See screenshots of errors thrown when running a sync with `wash` as the active deployment:
Screenshot 2023-02-06 at 16 15 33
Screenshot 2023-02-06 at 16 14 57

Should those values be populated by the default deployment config values from `DEPLOYMENT_CONFIG_EXAMPLE_DEFAULTS` in `packages/data-models/deployment.model.ts`? Or should consumers of the `activeDelpyment` config merge in the defaults after getting `ActiveDeployment.get()`?

**Comment by chrismclarke** [comment]
> when setting a deployment to active and compiling the activeDeployment.json file, default values are not set for local_drive, app_data, translations and workflows.

Nice catch @jfmcquade - yes each deployment should inherit the defaults from [DEPLOYMENT_CONFIG_EXAMPLE_DEFAULTS](https://github.com/IDEMSInternational/parenting-app-ui/blob/refactor/app-sync-scripts/packages/data-models/deployment.model.ts#L74-L75) but that only happens when config defined when defined in the newer syntax
```
import { generateDeploymentConfig } from "scripts";
const config = generateDeploymentConfig("plh");
```

I can see WASH still has the older syntax
```
import { getDefaultAppConfig } from "data-models/appConfig";
const app_config = getDefaultAppConfig();
```
So will update it now

File: src/app/shared/components/template/services/template-asset.service.ts [hunk 1]
```diff
@@ -55,15 +55,15 @@ export class TemplateAssetService extends AsyncServiceBase {
    * Fallsback to original path if does not exist
    */
   getThemeAssetPath(value: string) {
-    const themeFolderName = `theme_${this.themeService.getCurrentTheme()}`;
+    const themeFolderName = this.themeService.getCurrentTheme();
```
[the remark this answers is not in the document]
Comment (chrismclarke) [reply src/app/shared/components/template/services/template-asset.service.ts:58]: agreed, updated in fa03eb33e64332b195aa52f5c32648a81aa63499

**Comment by chrismclarke** [comment]
@jfmcquade 
I've just merged the pr to fix the wash deployment config. I realised the issue you flag still persists because the default value for the language codes was `null` and not `[]`, so I've added the extra conditional checks as well so hopefully should be all resolved now. Have requested re-review

**Review by jfmcquade — APPROVED** [review]
Looks good, the issue I flagged is fixed. The CLI experience of switching deployments and syncing the respective content is smooth

**Review by esmeetewinkel — CHANGES_REQUESTED** [review]
I found a bug, on the PLH TZ deployment the image assets don't seem to update correctly with the theme. 

To recreate: 
1. `yarn workflow deployment set plh_tz`
2. `yarn workflow clear_workflow_cache`
3. `yarn workflow sync`
4. `yarn start`
5. Navigate to http://localhost:4200/template/language_select and set language to English
6. Navigate to http://localhost:4200/template/skin_select, type the code `app_dev` and select `module`. This should set the skin to be the modular skin and the theme to be the professional theme. 
7. The homescreen now shows the professional theme but with images from the default theme.  
image

N.B. the audio assets DO seem to change correctly when changing language between English and Swahili

**Comment by chrismclarke** [comment]
Thanks for catching that @esmeetewinkel , you are right that none of the theme assets were being copied over.

I've added a fix for the issue and additional test to track the behaviour in the future so hopefully should be working now as expected

image

**Pull request merged**